# IBM Developer Skills Network

# Build a Deep Learning Regression Model - Assigment

## Import the required libraries

In [1]:
# for data import and clean
import pandas as pd  
import numpy as np

# for train test split
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# for building neural network
import keras
from keras.models import Sequential
from keras.layers import Dense

In [2]:
## Download and Clean Dataset
concrete_data = pd.read_csv('https://raw.githubusercontent.com/sonpn82/Introduction-to-deep-learning-and-neural-network/master/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [5]:
concrete_data.shape

(1030, 9)

## Check number of missing value

In [6]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

## Split data into predictors and target

In [3]:
# The target variable is the concrete sample strength
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
target = concrete_data['Strength']

## Build the neural network

In [5]:
# number of predictors for buidling neural network
n_cols = predictors.shape[1]
n_cols

8

In [6]:
# 1 hidden layer of 10 nodes, using ReLU activation function
# Use adam optimizer and mean squared error as loss function
def regression_model(hid_layer_no=1, hid_node_no=10, act_function='relu', opt_method='adam', loss_method='mean_squared_error'):
    # create the model
    model = Sequential()

    # create the input layer and the 1st hidden layer
    model.add(Dense(hid_node_no[0], activation=act_function, input_shape=(n_cols,))) 

    # add more hidden layer 
    if hid_layer_no > 1:
        for i in range(1, hid_layer_no):
            # add the hidden layer
            model.add(Dense(hid_node_no[i], activation=act_function)) 
    
    # add the output layer
    model.add(Dense(1))

    # compile model
    model.compile(optimizer=opt_method, loss=loss_method) 
    return model

## Train and Test the Network

In [7]:
# Set the parameter
hid_layer_no = 1
hid_node_no = [10]
act_function= 'relu'
opt_method = 'adam'
loss_method = 'mean_squared_error'

model = regression_model(hid_layer_no, hid_node_no, act_function, opt_method, loss_method)

In [8]:
mse = []
for i in range(50):
    # Train test split
    predict_train, predict_test, target_train, target_test = train_test_split(predictors, target, test_size=0.3, random_state=i)

    # Fit the model
    model.fit(predict_train, target_train, validation_split=0.3,epochs=50,verbose=0)

    # predict the test result
    target_predict = model.predict(predict_test)

    # mean square error calculation
    mse.append(mean_squared_error(target_test, target_predict))

print('Finish!')

Finish!


In [12]:
# Get mean and standard deviation
import statistics
print('Mean squared error: Average=', statistics.mean(mse))
print('Mean squared error: Standard Deviation=', statistics.pstdev(mse))

Mean squared error: Average= 70.42654546851655
Mean squared error: Standard Deviation= 70.8034036908935


## Normalize the data

In [14]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [15]:
mse_norm = []
for i in range(50):
    # Train test split
    predict_train, predict_test, target_train, target_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)

    # Fit the model
    model.fit(predict_train, target_train, validation_split=0.3,epochs=50,verbose=0)

    # predict the test result
    target_predict = model.predict(predict_test)

    # mean square error calculation
    mse_norm.append(mean_squared_error(target_test, target_predict))

print('Finish!')

Finish!


In [16]:
# Get mean and standard deviation
import statistics
print('Mean squared error: Average=', statistics.mean(mse_norm))
print('Mean squared error: Standard Deviation=', statistics.pstdev(mse_norm))

Mean squared error: Average= 72.1002747983437
Mean squared error: Standard Deviation= 77.31183496865734


### Mean square error is not much different between original data (70.4) and normalized data (72.1)

In [17]:
mse_norm_100 = []
for i in range(50):
    # Train test split
    predict_train, predict_test, target_train, target_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)

    # Fit the model
    model.fit(predict_train, target_train, validation_split=0.3,epochs=100,verbose=0)

    # predict the test result
    target_predict = model.predict(predict_test)

    # mean square error calculation
    mse_norm_100.append(mean_squared_error(target_test, target_predict))

print('Finish!')

Finish!


In [18]:
# Get mean and standard deviation
import statistics
print('Mean squared error: Average=', statistics.mean(mse_norm_100))
print('Mean squared error: Standard Deviation=', statistics.pstdev(mse_norm_100))

Mean squared error: Average= 34.815191631755575
Mean squared error: Standard Deviation= 3.289390824167354


### Mean square error of 100 epochs model is much lower than 50 epochs model (34.8 vs 72.1)

## Change the neural network to 3 hidden layers

In [20]:
# Set the parameter
hid_layer_no = 3
hid_node_no = [10,10,10]
act_function= 'relu'
opt_method = 'adam'
loss_method = 'mean_squared_error'

model_3layer = regression_model(hid_layer_no, hid_node_no, act_function, opt_method, loss_method)

In [21]:
mse_norm_3layer = []
for i in range(50):
    # Train test split
    predict_train, predict_test, target_train, target_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)

    # Fit the model
    model_3layer.fit(predict_train, target_train, validation_split=0.3,epochs=50,verbose=0)

    # predict the test result
    target_predict = model_3layer.predict(predict_test)

    # mean square error calculation
    mse_norm_3layer.append(mean_squared_error(target_test, target_predict))

print('Finish!')

Finish!


In [22]:
# Get mean and standard deviation
import statistics
print('Mean squared error: Average=', statistics.mean(mse_norm_3layer))
print('Mean squared error: Standard Deviation=', statistics.pstdev(mse_norm_3layer))

Mean squared error: Average= 51.32451736831406
Mean squared error: Standard Deviation= 32.17597857387679


### Mean squared error of 3 layers model is lower than 1 layer model (51.3 vs 72.1)